### PR TITLE
chore(atomix): add checksum to append request

### LINKED ISSUE #4736: Add checksum to AppendEntries request
**Description**

We should add checksums to AppendEntries requests in order to ensure that followers don't write corrupted entries.

### PR DESCRIPTION
**Pull request by miguelpires**
## Description

Adds a checksum to the AppendRequest that is replicated by the leader and then checked by the followers.

## Related issues

closes #4736

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [x] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

File: atomix/cluster/src/main/java/io/atomix/raft/roles/PassiveRole.java [hunk 1]
```diff
@@ -618,6 +648,23 @@ protected void appendEntries(
...
+  private boolean validChecksum(final AppendRequest request, final int index) {
+    if (request.checksums() == null) {
+      return true;
+    }
+
+    final long checksum = request.checksums().get(index);
+    final RaftLogEntry entry = request.entries().get(index);
+
+    buffer.clear();
+    namespace.serialize(entry, buffer);
```
Comment (deepthidevaki) [inline atomix/cluster/src/main/java/io/atomix/raft/roles/PassiveRole.java:660]: We are serializing the entry and calculating the checksums here and again in journal::append. Can we not do that?

**Review by deepthidevaki — CHANGES_REQUESTED** [review]
Thanks @MiguelPires I have some comments on the approach:
1. On the receiver we are serializing and calculating checksums twice - in the Raft roles and again in the journal append. 
2. On the reader journal is calculating the checksum and the raft leader only reads it. On the receiver raft is calculating and verifying the checksum. May be we can refactor it in such a way that only one module needs to know how to calculate a checksum?
Wdyt?

File: atomix/cluster/src/main/java/io/atomix/raft/roles/PassiveRole.java [hunk 2]
```diff
@@ -618,6 +648,23 @@ protected void appendEntries(
     succeedAppend(lastLogIndex, future);
   }
 
+  private boolean validChecksum(final AppendRequest request, final int index) {
+    if (request.checksums() == null) {
```
[the remark this answers is not in the document]
Comment (MiguelPires) [reply atomix/cluster/src/main/java/io/atomix/raft/roles/PassiveRole.java:652]: Yes

File: atomix/cluster/src/main/java/io/atomix/raft/roles/PassiveRole.java (hunk 1 above)
Comment (MiguelPires) [reply atomix/cluster/src/main/java/io/atomix/raft/roles/PassiveRole.java:660]: I didn't see a way of doing that without refactoring the journal. We'd have to have slightly different journal implementations for the leader and follower

Comment (deepthidevaki) [reply atomix/cluster/src/main/java/io/atomix/raft/roles/PassiveRole.java:660]: Ok. Let's leave it as it is. Can you create a follow up issue for it? Did you run a benchmark with this changes - does having to serialize it twice has any impact on the performance?

Comment (MiguelPires) [reply atomix/cluster/src/main/java/io/atomix/raft/roles/PassiveRole.java:660]: Yes, it's in the mp-4736 namespace. It's been running for 10 days without issues. The average for the last 7 days is 142 processes completed per second. I do see the checksum validation in the profiling at ~1% due to the serialization

**Review by deepthidevaki — APPROVED** [review]
:100:

**Comment by MiguelPires** [comment]
Thanks for the review @deepthidevaki. I reworked this PR to not do the serialization twice. The solution simply passes a validator when appending, which the writer uses to check that the entry is valid. This is not an ideal design but given the current state it's the simplest and clearest. I tried refactoring the checksum parts into a new entity but it ended up being more confusing since there's no clear abstraction (it doesn't just depend on whether it's leader or follower, there's more). So I went with the simplest solution

File: atomix/cluster/src/main/java/io/atomix/raft/roles/PassiveRole.java [hunk 3]
```diff
@@ -706,18 +747,25 @@ private boolean replaceExistingEntry(
...
+
       log.trace("Appended {}", indexed);
       raft.getReplicationMetrics().setAppendIndex(indexed.index());
     } catch (final StorageException.TooLarge e) {
       log.warn(
           "Entry size exceeds maximum allowed bytes. Ensure Raft storage configuration is consistent on all nodes!");
       return false;
     } catch (final StorageException.OutOfDiskSpace e) {
-      log.trace("Append failed: {}", e);
+      log.trace("Append failed: ", e);
```
Comment (deepthidevaki) [inline atomix/cluster/src/main/java/io/atomix/raft/roles/PassiveRole.java:768]: Must catch `StorageException.InvalidChecksum`

File: atomix/cluster/src/main/java/io/atomix/raft/roles/PassiveRole.java [hunk 4]
```diff
@@ -580,13 +611,20 @@ protected void appendEntries(
       }
 
       // Iterate through entries and append them.
-      for (final RaftLogEntry entry : request.entries()) {
+      for (int i = 0; i < request.entries().size(); ++i) {
         final long index = ++lastLogIndex;
+        final RaftLogEntry entry = request.entries().get(i);
+
+        AtomicInteger checksum = null;
```
Comment (deepthidevaki) [inline atomix/cluster/src/main/java/io/atomix/raft/roles/PassiveRole.java:618]: Why is it AtomicInteger? There is no concurrent access to it.

File: atomix/cluster/src/main/java/io/atomix/raft/roles/PassiveRole.java [hunk 5]
```diff
@@ -437,11 +440,39 @@ private void abortPendingSnapshots() {
...
     appendEntries(request, future);
     return future;
   }
 
+  private boolean checkChecksums(
+      final AppendRequest request, final CompletableFuture<AppendResponse> future) {
+    if (request.checksums() != null) {
+      boolean failed = false;
+      if (request.entries().size() != request.checksums().size()) {
+        log.debug("Rejected {}: expected the same number of checksums as entries", request);
```
Comment (deepthidevaki) [inline atomix/cluster/src/main/java/io/atomix/raft/roles/PassiveRole.java:457]: Also add request size and checksum size to the log message.

File: atomix/cluster/src/test/java/io/atomix/raft/roles/PassiveRoleTest.java [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+
+    // when
+    final AppendResponse response = role.handleAppend(request).join();
+
+    // then
+    assertThat(response.succeeded()).isFalse();
+  }
+
+  @Test
+  public void shouldCreateValidatorForEntry() {
```
Comment (deepthidevaki) [inline atomix/cluster/src/test/java/io/atomix/raft/roles/PassiveRoleTest.java:104]: Please rename this function.

File: atomix/storage/src/main/java/io/atomix/storage/journal/FileChannelJournalSegmentWriter.java [hunk 7]
```diff
@@ -94,62 +95,85 @@ public long getNextIndex() {
...
+  public <T extends E> Indexed<T> append(final T entry, final int checksum) {
+    return append(entry, true, checksum);
+  }
+
   @Override
   public <T extends E> Indexed<T> append(final T entry) {
+    return append(entry, false, -1);
+  }
+
+  public <T extends E> Indexed<T> append(
```
Comment (deepthidevaki) [inline atomix/storage/src/main/java/io/atomix/storage/journal/FileChannelJournalSegmentWriter.java:108]:
```suggestion
  private <T extends E> Indexed<T> append(
```

File: atomix/storage/src/main/java/io/atomix/storage/journal/MappedJournalSegmentWriter.java [hunk 8]
```diff
@@ -97,56 +98,75 @@ public long getNextIndex() {
...
+    return append(entry, false, -1);
+  }
+
+  @Override
+  public <T extends E> Indexed<T> append(final T entry, final int checksum) {
+    return append(entry, true, checksum);
+  }
 
-    // Serialize the entry.
+  public <T extends E> Indexed<T> append(
```
Comment (deepthidevaki) [inline atomix/storage/src/main/java/io/atomix/storage/journal/MappedJournalSegmentWriter.java:109]:
```suggestion
  private <T extends E> Indexed<T> append(
```

File: atomix/cluster/src/test/java/io/atomix/raft/roles/PassiveRoleTest.java [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import java.util.zip.CRC32;
+import java.util.zip.Checksum;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.Timeout;
+
+public class PassiveRoleTest {
```
Comment (deepthidevaki) [inline atomix/cluster/src/test/java/io/atomix/raft/roles/PassiveRoleTest.java:54]: We are missing a test to verify what happens when checksum is invalid.

**Review by deepthidevaki — CHANGES_REQUESTED** [review]
Looks good. Some small comments.

File: atomix/cluster/src/main/java/io/atomix/raft/roles/PassiveRole.java (hunk 5 above)
Comment (MiguelPires) [reply atomix/cluster/src/main/java/io/atomix/raft/roles/PassiveRole.java:457]: That's what the request prints 
```
@Override
  public String toString() {
    return toStringHelper(this)
        .add("term", term)
        .add("leader", leader)
        .add("prevLogIndex", prevLogIndex)
        .add("prevLogTerm", prevLogTerm)
        .add("entries", entries.size())
        .add("checksums", checksums.size())
        .add("commitIndex", commitIndex)
        .toString();
  }
```

File: atomix/cluster/src/test/java/io/atomix/raft/roles/PassiveRoleTest.java (hunk 9 above)
Comment (MiguelPires) [reply atomix/cluster/src/test/java/io/atomix/raft/roles/PassiveRoleTest.java:54]: That's not checked in the PassiveRole. That test is in the JournalSegmentWriterTest

**Review by deepthidevaki — APPROVED** [review]
:+1:

**Pull request merged**